packages/test/local-server-stress-tests/src/stressDataObject.ts

```diff
@@ -0,0 +1,320 @@
+
+		const maybe: FluidObject<DefaultStressDataObject> | undefined =
+			await defaultDataStore.get();
+		assert(maybe.DefaultStressDataObject !== undefined, "must be DefaultStressDataObject");
+		return maybe.DefaultStressDataObject;
+	}
+
+	/**
+	 * this map is special, and doesn't participate in stress. it hold data
+	 * about the name of channels which have been created. these created channel
+	 * may or may not be attached and be available
+	 */
+	private channelNameMap: ISharedMap = makeUnreachableCodePathProxy("channelNameMap");
+	protected async initializingFirstTime(props?: any): Promise<void> {
+		this.channelNameMap = SharedMap.create(this.runtime, "channelNameMap");
+		this.channelNameMap.bindToContext();
+		this.channelNameMap.set("root", this.root.attributes.type);
+	}
+
+	public async getChannels() {
+		const channels: IChannel[] = [];
+		for (const [name] of this.channelNameMap.entries()) {
+			// similar to container objects, the entries in this map
+			// can appear before the underlying channel is attached,
+			// so getting the channel can fail, and we need to try
+			// to get all channel each time, as we have no way to
+			// observer when a channel moves from detached to attached,
+			// especially on remove clients/
+			const channel = await this.runtime.getChannel(name).catch(() => undefined);
+			if (channel !== undefined) {
+				channels.push(channel);
+			}
+		}
+		return channels;
+	}
+
+	protected async hasInitialized(): Promise<void> {
+		this.defaultStressObject = await this.getDefaultStressDataObject();
+
+		this.channelNameMap = (await this.runtime.getChannel(
+			"channelNameMap",
+		)) as any as ISharedMap;
+	}
+
+	public get attached() {
+		return this.runtime.attachState === AttachState.Attached;
+	}
+
+	public uploadBlob(tag: `blob-${number}`, contents: string) {
+		void this.runtime.uploadBlob(stringToBuffer(contents, "utf-8")).then((handle) =>
+			this.defaultStressObject.registerLocallyCreatedObject({
+				type: "newBlob",
+				handle,
+				tag,
+			}),
+		);
+	}
+
+	public createChannel(tag: `channel-${number}`, type: string) {
+		this.runtime.createChannel(tag, type);
+		this.channelNameMap.set(tag, type);
+	}
+
+	public createDataStore(tag: `datastore-${number}`, asChild: boolean) {
+		void this.context.containerRuntime
+			.createDataStore(
+				asChild
+					? [...this.context.packagePath, StressDataObject.factory.value.type]
+					: StressDataObject.factory.value.type,
+			)
+			.then(async (dataStore) => {
+				const maybe: FluidObject<StressDataObject> | undefined =
+					await dataStore.entryPoint.get();
+				assert(maybe?.StressDataObject !== undefined, "must be stressDataObject");
+				this.defaultStressObject.registerLocallyCreatedObject({
+					type: "stressDataObject",
+					handle: dataStore.entryPoint,
+					tag,
+					stressDataObject: maybe.StressDataObject,
+				});
+			});
+	}
+}
+export type ContainerObjects =
+	| { type: "newBlob"; handle: IFluidHandle; tag: `blob-${number}` }
```

Comment:
That's clever (but maybe also a little spooky?), I didn't know you could use template string-like syntax in types.

Reply:
https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html